### PR TITLE
Fix bool.TryParse/Format on big-endian systems

### PR DESCRIPTION
**Pull request by uweigand**
* Fix regression after https://github.com/dotnet/runtime/pull/64782

CC @stephentoub

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Boolean.cs [hunk 1]
```diff
@@ -110,7 +117,14 @@ public bool TryFormat(Span<char> destination, out int charsWritten)
...
                 {
-                    BinaryPrimitives.WriteUInt64LittleEndian(MemoryMarshal.AsBytes(destination), 0x73006C00610046); // "Fals"
+                    if (BitConverter.IsLittleEndian)
+                    {
+                        BinaryPrimitives.WriteUInt64LittleEndian(MemoryMarshal.AsBytes(destination), 0x73006C00610046); // "Fals"
+                    }
+                    else
+                    {
+                        BinaryPrimitives.WriteUInt64BigEndian(MemoryMarshal.AsBytes(destination), 0x460061006C0073); // "Fals"
+                    }
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Boolean.cs:127]: I don't understand, wasn't `.WriteUInt64LittleEndian` supposed to handle endianess via ReverseEndianness() for value?
```

        /// <summary>
        /// Write a UInt64 into a span of bytes as little endian.
        /// </summary>
        [CLSCompliant(false)]
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static void WriteUInt64LittleEndian(Span<byte> destination, ulong value)
        {
            if (!BitConverter.IsLittleEndian)
            {
                value = ReverseEndianness(value);
            }
            MemoryMarshal.Write(destination, ref value);
        }
```

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Boolean.cs:127]: if it wasn't then you can re-write code to just `MemoryMarshal.Write(destination, 0x73006C00610046);` I guess

Comment (uweigand) [reply src/libraries/System.Private.CoreLib/src/System/Boolean.cs:127]: The problem is that you need to distinguish between swapping the order of the *characters* and the order of the *bytes* within one two-byte character.   See https://github.com/dotnet/runtime/pull/64782#issuecomment-1033785023 for an example.  There's no single store that does the correct thing for both of these simultaneously on both big- and little-endian platforms.

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Boolean.cs:127]: I see, thanks for explanation! anyway maybe it's still better to use 
```
MemoryMarshal.Write(destination, IsLittleEndian ? 0x73006C00610046 : ...)
```
? 
less verbose IMO

Comment (uweigand) [reply src/libraries/System.Private.CoreLib/src/System/Boolean.cs:127]: That would work as well.   I can try this out if you prefer.

Comment (uweigand) [reply src/libraries/System.Private.CoreLib/src/System/Boolean.cs]: Patch updated accordingly.

**Review by EgorBo — APPROVED** [review]

**Review by jkotas — APPROVED** [review]

**Review by stephentoub — COMMENTED** [review]
Thanks for fixing.

Are we running the regex tests on the s390x?  I would expect it's going to have similar problems.

**Comment by stephentoub** [comment]
> Those seem to be all good now ...

Thanks for looking.  I think there's actually still an issue there, but it's masked by limitations in our test infrastructure.  .NET 7 includes a source generator for regex, where the generator will emit into the .dll the code for the regex implementation.  I expect if you were to take a .dll compiled on a little-endian machine and run it on the s390x, or vice versa, you'd see operations fail.

Can you try compiling this program using a .NET 7 SDK on a little-endian machine:
```C#
using System.Text.RegularExpressions;

partial class Program
{
    public static void Main() => Console.WriteLine(Example().IsMatch(@"1abcd"));

    [RegexGenerator(@"\dabcd")]
    public static partial Regex Example();
}
```
and then running the resulting program on the big-endian one?  It should print true, but I'm betting it's going to print false.

I'll prepare a fix for it, but I'm going to need your help in validating its correctness.

**Comment by uweigand** [comment]
> Thanks for looking. I think there's actually still an issue there, but it's masked by limitations in our test infrastructure. .NET 7 includes a source generator for regex, where the generator will emit into the .dll the code for the regex implementation. I expect if you were to take a .dll compiled on a little-endian machine and run it on the s390x, or vice versa, you'd see operations fail.

I see.

> It should print true, but I'm betting it's going to print false.

Yes, this is indeed what happens.

> I'll prepare a fix for it, but I'm going to need your help in validating its correctness.

Sure, happy to help!   Thanks for thinking of this.

**Comment by uweigand** [comment]
> > > I'll prepare a fix for it, but I'm going to need your help in validating its correctness.
> 
> > Sure, happy to help! Thanks for thinking of this.
> 
> I ended up just deleting the offending code instead.

I repeated the test above using today's snapshot (dotnet-sdk-7.0.100-preview.2.22118.2-linux-x64.tar.gz) to build the regex test case on Intel, and can confirm that the resulting assembly now outputs "True" when run on s390x.

Thanks!

**Comment by stephentoub** [comment]
Thanks for confirming, @uweigand.

**Pull request merged**